design/gas/Design_gas.md

```diff
@@ -0,0 +1,89 @@
+2. Bridged gas should automatically deposit into the respective addresses.
+3. Gas-consuming transactions should include a layer 1 cost component for calldata publishing and a layer 2 component that covers execution costs and prevents endless transaction execution.
+4. L2 gas price should be configurable. Furthermore it should support being free whilst still not allowing non halting execution.
+5. The gas mechanics need to prevent denial of service attacks.
+6. EIP-1559 transactions should have their base fee + tip deposited to the sequencer without any ETH being burned.
+
+## Gas Cost Display
+
+Metamask can disassemble the cost components for layer 1 and layer 2 for specific transactions on Optimism, such as token balance approvals. However, it remains unclear how they achieve this as it's neither fully documented nor supports all transactions.
+
+
+## Design for Gas Deposit
+
+Given that Ethereum as a currency operates on the layer 1 protocol, it needs to be bridged to Obscuro for usage. This transition mechanism will be facilitated by the standard Obscuro bridge. The deposited value will produce a cross-chain message indicating the Ethereum recipient through the smart contract.
+
+```EthereumDeposited(address receiver, uint256 amount)```
+
+The enclave will automatically pick up this message, along with other cross-chain messages, but it will be treated specially. Upon detection, the enclave will increase the balance of the receiver account by the amount specified in the message. This automatic relaying is needed in order to avoid the chicken and egg problem one would have initially where no relayer has any gas to relay the gas deposit messages. It would also ensure relayers who run out of gas can easily recharge without relying on other relayers if we were to use a different bootstrap mechanism. Note that those deposits will not call the fallback function of the address if it is a smart contract and any deposits to such contracts would later on need a message relayed to verify the message as they would normally.
+
+## Gas Estimation
+
+When an RPC call for gas estimation is made to Obscuro, the returned estimate should include the expected layer 1 calldata cost alongside the usual execution-based gas estimation. This can be achieved by taking an estimation for layer 1 minFee per gas, determining the calldata cost of the transaction, and adding this to the standard gas estimate.
+
+`Gas estimate = L1_gas_fee * calldata_gas + estimated_l2_gas`
+
+## Gas Expenditure
+
+Before executing a transaction at Obscuro, we must automatically deduct the layer 1 costs for the transaction from the sender's balance. This can be done directly through the stateDB before processing. Ideally, the remaining balance will be refunded, simulating a scenario where the gas limit has been spent. However, potential overruns of execution costs could result in transactions exceeding their set gas limit.
```

Comment:
Why `deduct the layer 1 costs for the transaction from the sender's balance` ?

What if the gas price was bumped for a given transaction? Such that :
- `gas_price = l2_current_gasPrice + projected_number_txs_in_a_rollup / l1_projected_gasPrice`

Reply:
This calculation makes it so transactions that spend more gas on the L2 to pay for more L1 fees, while value transfers will pay little in L1 fees. It's not really a competitive solution with other rollups. We'd have some users subsidizing others.

Reply:
I might be missing somthing, can you explain a bit more about how is one user subsidizing others? 
The lowest possible transaction is a value transfer, which has a fixed L2 gas cos, which will potentially be a fixed L1 Rollup cost, so it should pay a smaller L1 fee than a contract creation or other transactions right ?

Reply:
No, because the gas limit of that said transaction is higher than the value transfer.
So let's say there's a Quirky Tx that does a call on a contract which does a for loop (spins a while).

gasPrice: ( L2Gasprice + L1Gasprice)
Tx A: Value transfer - tx rollup size 10 - gas limit 21 000 - Final gas cost 21 000 * gasPrice
Tx B: Deploy Contract - tx rollup size 3000 - gas limit 120 000 - Final gas cost 120 000 * gasPrice
Tx C: Quirky Tx - tx rollup size 40 - gas limit 500 000 - Final gas cost 500 000 * gasPrice

The rollup size does not matter for the gas price nor for the final gas paid by the user.
Saying that tx A and C are subsidizing Tx B doesn't make a lot of sense because they are paying network congestion on the L2.

That being said, if you want to figure out a way that you can adjust the `L1GasPrice` component based on the potential tx rollup size, I think that's a good idea, but perhaps not needed for a first iteration imo.